### PR TITLE
BLOG-124 | How to setup CI/CD system for next.js with Docker on gitlab-ci, travis-ci

### PR DESCRIPTION
**Pull request by hoatle**

### CODE REVIEW

File: source/_posts/2017-05-09-how-to-set-up-ci-cd-system-for-next.js-with-docker-on-gitlab-ci-travis-ci-heroku-and-kubernetes.markdown [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+
+`nextjs-hello-world` is the simplest seed code for Next.js applications with Docker workflow, CI/CD system:
+
+- CI/CD with gitlab-ci: https://gitlab.com/acme101/nextjs-hello-world/pipelines
+
+- CI/CD with travis-ci: https://travis-ci.org/acme101/nextjs-hello-world/builds
+
+- Auto deployment to Heroku: https://acme-nextjs-staging.herokuapp.com/
+
+- Auto deployment to GKE (Kubernetes) with terapp.com domain A record: https://acme-nextjs-staging.terapp.com/
```
Comment (hoavt) [inline source/_posts/2017-05-09-how-to-set-up-ci-cd-system-for-next.js-with-docker-on-gitlab-ci-travis-ci-heroku-and-kubernetes.markdown]: terapp.com domain A record -->don't understand

File: source/_posts/2017-05-09-how-to-set-up-ci-cd-system-for-next.js-with-docker-on-gitlab-ci-travis-ci-heroku-and-kubernetes.markdown [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+
+To setup the project on local development:
```
Comment (hoavt) [inline source/_posts/2017-05-09-how-to-set-up-ci-cd-system-for-next.js-with-docker-on-gitlab-ci-travis-ci-heroku-and-kubernetes.markdown]: --> To set up

File: source/_posts/2017-05-09-how-to-set-up-ci-cd-system-for-next.js-with-docker-on-gitlab-ci-travis-ci-heroku-and-kubernetes.markdown [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+consistent behaviors between all developers and production deployments.
+
+And Docker helps us with that to create a consistent build-time and run-time environment for all.
+
+Usually, there are 3 modes on local dev for our workflow:
+
+- Dev Mode: developers work on this for new changes, this usually contains development dependencies.
+
+- Prod Mode: developers need to make sure that production Docker image should work on local dev.
+  This production Docker image is different from the one from dev mode is that it will contain only
```
Comment (hoavt) [inline source/_posts/2017-05-09-how-to-set-up-ci-cd-system-for-next.js-with-docker-on-gitlab-ci-travis-ci-heroku-and-kubernetes.markdown]: --> This production Docker image which is different from the one from dev mode will contain only

File: source/_posts/2017-05-09-how-to-set-up-ci-cd-system-for-next.js-with-docker-on-gitlab-ci-travis-ci-heroku-and-kubernetes.markdown [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+  the production dependencies, the runtime environment only.
```
Comment (hoavt) [inline source/_posts/2017-05-09-how-to-set-up-ci-cd-system-for-next.js-with-docker-on-gitlab-ci-travis-ci-heroku-and-kubernetes.markdown]: , and the runtime...

File: source/_posts/2017-05-09-how-to-set-up-ci-cd-system-for-next.js-with-docker-on-gitlab-ci-travis-ci-heroku-and-kubernetes.markdown [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+
+  If developers can only make it work on dev mode, prod mode can break. If prod mode breaks, developers can
+ check and fix it on local dev. This is really conveninent and time saving.
+
+- Prod Review Mode: we should review the work from others and this mode help us for faster reviewing
+  process. Basically, everyone's work branch will have the corresponding production Docker image that
+  we can review it right away on our local dev environment. We don't have to checkout the codes to start
+  reviewing.
+
+  This is helpful for us to setup CI/CD system for reviewing process later: when a pull request is sent,
```
Comment (hoavt) [inline source/_posts/2017-05-09-how-to-set-up-ci-cd-system-for-next.js-with-docker-on-gitlab-ci-travis-ci-heroku-and-kubernetes.markdown]: to set up

File: source/_posts/2017-05-09-how-to-set-up-ci-cd-system-for-next.js-with-docker-on-gitlab-ci-travis-ci-heroku-and-kubernetes.markdown [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+  the CI/CD system should deploy it right away for QA to validate, for example.
+
+
+More details how to use this modes below:
```
Comment (hoavt) [inline source/_posts/2017-05-09-how-to-set-up-ci-cd-system-for-next.js-with-docker-on-gitlab-ci-travis-ci-heroku-and-kubernetes.markdown]: The following is the more details about how to use these modes:

**Comment by hoatle** [comment]
@hoavt feedback applied.

**Comment by hoatle** [comment]
@hoavt for A record domain, see: https://pressable.com/blog/2014/12/23/dns-record-types-explained/

**Pull request merged**